sys/dev/cxgbe/t4_ioctl.h

```diff
@@ -401,7 +401,14 @@ struct offload_rule {
  */
 struct t4_offload_policy {
 	uint32_t nrules;
-	struct offload_rule *rule;
```

Comment:
Can't this be __kerncap and we update the non-cap uses of rule?

Reply:
set_offload_policy has:
```C
        op->rule = malloc(len, M_CXGBE, M_ZERO | M_WAITOK);
        rc = copyin(uop->user_rule, op->rule, len);
```
and turning a capability back into a pointer for the destination is a little scary (if you get it right it's fine, but if you get it wrong...).

Reply:
Maybe this then:
```
	void *tmp = malloc(len, M_CXGBE, M_ZERO | M_WAITOK);
	rc = copyin(uop->rule, tmp, len);
	op->rule = PTR2CAP(tmp);
```
And we should probably add a PTR2CAP() version that allows passing a length.

Alternatively a copyin variant that expects a capability destination?

Reply:
You then need a `__cheri_fromcap` cast when you come to free the thing, and change all the uses of the kernel copy of the structure to have `__capability`. It's doable, it's just a bigger diff, and is also less in the spirit of a hybrid kernel.

Reply:
malloc.h defines `free_c`, so that can be used to avoid the cast. Not sure if the diff is any bigger since you no longer need the  rule->user_rule changes.

Reply:
I don't like the asymmetry of `malloc` and `free_c` though. And the diff is definitely smaller, `user_rule` is only needed in one place.

Reply:
But then you need the anonymous union. How about we add a `copyin_c()`, then we can use malloc_c + free_c?

Reply:
We do this trick all over the place already in CAM code.

Reply:
It's the easiest way to make things work, and also helps you not screw up and confuse kernel/user pointers. Arguably it's how it should be done even in a non-CHERI or purecap kernel in order to be able to clearly annotate what's a user pointer and what's a kernel pointer so you can statically check they never get used in the wrong way.

Reply:
Well you can still write the user pointer and read it as a kernel pointer? But I agree that you do at least show what's intended.